File: framework/security/src/main/java/top/bettercode/simpleframework/security/repository/InMemoryApiTokenRepository.java
package top.bettercode.simpleframework.security.repository;

import java.util.Map;
import top.bettercode.simpleframework.security.ApiToken;

/**
 * @author Peter Wu
 */
public class InMemoryApiTokenRepository implements ApiTokenRepository {

  private final Map<String, ApiToken> tokenMap;
  private final Map<String, String> accessTokenMap;
  private final Map<String, String> refreshTokenMap;

  public InMemoryApiTokenRepository(
      Map<String, ApiToken> tokenMap,
      Map<String, String> accessTokenMap,
      Map<String, String> refreshTokenMap) {
    this.tokenMap = tokenMap;
    this.accessTokenMap = accessTokenMap;
    this.refreshTokenMap = refreshTokenMap;
  }

  @Override
  public void save(ApiToken apiToken) {
    String scope = apiToken.getScope();
    String username = apiToken.getUsername();
    String id = scope + ":" + username;
    remove(scope, username);
    accessTokenMap.put(apiToken.getAccessToken().getTokenValue(), id);
    refreshTokenMap.put(apiToken.getRefreshToken().getTokenValue(), id);
    tokenMap.put(id, apiToken);
  }

  @Override
  public void remove(ApiToken apiToken) {
    String scope = apiToken.getScope();
    String username = apiToken.getUsername();
    String id = scope + ":" + username;
    accessTokenMap.remove(apiToken.getAccessToken().getTokenValue());
    refreshTokenMap.remove(apiToken.getRefreshToken().getTokenValue());
    tokenMap.remove(id);
  }

  @Override
  public void remove(String scope, String username) {
    ApiToken authenticationToken = findByScopeAndUsername(scope, username);
    if (authenticationToken != null) {
      remove(authenticationToken);
    }
  }

  @Override
  public ApiToken findByScopeAndUsername(String scope, String username) {
    String id = scope + ":" + username;
    return tokenMap.get(id);
  }

  @Override
  public ApiToken findByAccessToken(String accessToken) {
    String id = accessTokenMap.get(accessToken);
    if (id != null) {
      return tokenMap.get(id);
    }
    return null;
  }

  @Override
  public ApiToken findByRefreshToken(String refreshToken) {
    String id = refreshTokenMap.get(refreshToken);
    if (id != null) {
      return tokenMap.get(id);
    }
    return null;
  }

}

File: security/src/main/kotlin/top/bettercode/summer/security/repository/InMemoryApiTokenRepository.kt
package top.bettercode.summer.security.repository

import top.bettercode.summer.security.token.ApiToken

/**
 * @author Peter Wu
 */
class InMemoryApiTokenRepository(
        private val tokenMap: MutableMap<String, ApiToken?>,
        private val accessTokenMap: MutableMap<String?, String>,
        private val refreshTokenMap: MutableMap<String?, String>
) : ApiTokenRepository {
    override fun save(apiToken: ApiToken) {
        val scope = apiToken.scope
        val username = apiToken.username
        val id = "$scope:$username"
        remove(scope, username)
        accessTokenMap[apiToken.accessToken.tokenValue] = id
        refreshTokenMap[apiToken.refreshToken.tokenValue] = id
        tokenMap[id] = apiToken
    }

    override fun remove(apiToken: ApiToken) {
        val scope = apiToken.scope
        val username = apiToken.username
        val id = "$scope:$username"
        accessTokenMap.remove(apiToken.accessToken.tokenValue)
        refreshTokenMap.remove(apiToken.refreshToken.tokenValue)
        tokenMap.remove(id)
    }

    override fun remove(scope: String, username: String) {
        val id = "$scope:$username"
        val authenticationToken = tokenMap[id]
        authenticationToken?.let { remove(it) }
    }

    override fun remove(scope: String, usernames: List<String>) {
        usernames.forEach { remove(scope, it) }
    }

    override fun findByScopeAndUsername(scope: String, username: String): ApiToken? {
        val id = "$scope:$username"
        return tokenMap[id]
    }

    override fun findByAccessToken(accessToken: String): ApiToken? {
        val id = accessTokenMap[accessToken]
        return if (id != null) {
            tokenMap[id]
        } else null
    }

    override fun findByRefreshToken(refreshToken: String): ApiToken? {
        val id = refreshTokenMap[refreshToken]
        return if (id != null) {
            tokenMap[id]
        } else null
    }
}
